### PR TITLE
hdl._nir: add combinational cycle detection.

### LINKED ISSUE #704: Detect and reject netlists with combinatorial loops
_Originally posted by @whitequark in https://github.com/amaranth-lang/amaranth/issues/703#issuecomment-1104789392_

> In general, if the simulator doesn't converge, it means an issue with the input netlist. I'd say that the appropriate way to handle this is to reject such netlists before they can be simulated, since by the time the simulator is stuck, the original issue (a logic loop of some kind) becomes very hard to find.

&rArr; Questions

* How to detect such netlists : loops, ... ?
* Cases that can't be spotted ?
* Message to report the problem ?

### LINKED ISSUE #1143: infinite loop in NIR builder on a trivial comb loop
The following results in an infinite loop:

```
from amaranth import *
import amaranth.back.rtlil

m = Module()
a = Signal()
m.d.comb += a.eq(a)

amaranth.back.rtlil.convert(m, ports=[])
```

### PR DESCRIPTION
**Pull request by wanda-phi**
Fixes #704.
Fixes #1143.

### CODE REVIEW

File: tests/test_hdl_ir.py [hunk 1]
```diff
@@ -3542,3 +3542,16 @@ class MyEnum(enum.Enum, shape=unsigned(2)):
...
+class CycleTestCase(FHDLTestCase):
+    def test_cycle(self):
+        a = Signal()
+        b = Signal()
+        m = Module()
+        m.d.comb += [
+            a.eq(~b),
+            b.eq(~a),
+        ]
+        with self.assertRaises(CombinationalCycle):
```
Comment (whitequark) [inline tests/test_hdl_ir.py:3556]: This should probably assert on a message using `assertRaisesRegex` to skip the variable parts like the line numbers.

File: amaranth/hdl/_nir.py [hunk 2]
```diff
@@ -698,6 +804,10 @@ def __repr__(self):
         patterns = " ".join(self.patterns)
         return f"(matches {self.value} {patterns})"
 
+    def comb_edges_to(self, bit: int) -> Iterable[Net]:
```
Comment (whitequark) [inline amaranth/hdl/_nir.py:807]: The annotation on the base class is probably sufficient.

File: amaranth/hdl/_nir.py [hunk 3]
```diff
@@ -733,6 +843,11 @@ def resolve_nets(self, netlist: Netlist):
     def __repr__(self):
         return f"(priority_match {self.en} {self.inputs})"
 
+    def comb_edges_to(self, bit: int) -> Iterable[Net]:
```
Comment (whitequark) [inline amaranth/hdl/_nir.py:846]: Likewise.

File: amaranth/hdl/_nir.py [hunk 4]
```diff
@@ -405,16 +412,81 @@ def add_value_cell(self, width: int, cell):
...
+                        sig, bit = self.late_to_signal[net]
+                        e.path.append((sig, bit, sig.src_loc))
+                        raise
+                else:
+                    for src, src_loc in self.cells[net.cell].comb_edges_to(net.bit):
+                        try:
+                            traverse(src)
+                        except Cycle as e:
+                            e.path.append((self.cells[net.cell], net.bit, src_loc))
+                            raise
```
Comment (whitequark) [inline amaranth/hdl/_nir.py:462]: Please use a local return instead of a nonlocal return, since using a nonlocal return adds nothing.

**Review by whitequark — CHANGES_REQUESTED** [review]
See inline comments.

**Review by whitequark — APPROVED** [review]
LGTM, thanks!

**Comment by widlarizer** [comment]
Sweet!

**Pull request merged**